runner_service/utils.py

```diff
@@ -206,9 +207,14 @@ def timeout_handler():
             raise SSHTimeout
 
         socket.setdefaulttimeout(self.timeout)
-        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
         try:
-            s.connect((self.host, self.port))
+            if ipaddress.ip_address(self.host) == 4:
```

Comment:
Hmm, unless I'm missing something this will raise ValueError if you have FQDN specified in host. I think we should use something like this:
```
ipv6_addr = True
try:
    ipaddress.IPv6Address(self.host)
except ipaddress.AddressValueError:
    ipv6_addr = False

if ipv6_addr:
    s = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
    s.connect((self.host, self.port, 0, 0))
else:
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((self.host, self.port))
```

Reply:
Marcin is there any better solution how to support all possible combination of host specification (FQDN, IPv4 or IPv6)?

@tinez

Reply:
I think it's better to start off with [`socket.getaddrinfo`](https://docs.python.org/3.6/library/socket.html?highlight=socket%20connect#socket.getaddrinfo). The way it works, is it returns a list of 5-element tuples that can be used to pass to `socket.socket`.

- If you use it with FQDN it will return ipv6 and/or ipv4 addresses in the order that is determined by DNS resolving configuration in your OS i.e.:

```
>>> pp(socket.getaddrinfo("localhost", 22, 0, socket.SOCK_STREAM))
[(<AddressFamily.AF_INET6: 10>,
  <SocketKind.SOCK_STREAM: 1>,
  6,
  '',
  ('::1', 22, 0, 0)),
 (<AddressFamily.AF_INET: 2>,
  <SocketKind.SOCK_STREAM: 1>,
  6,
  '',
  ('127.0.0.1', 22))]
```
- if you use ipv4 it will limit the results to ipv4 addresses only, i.e.:
```
>>> pp(socket.getaddrinfo("127.0.0.1", 22, 0, socket.SOCK_STREAM))
[(<AddressFamily.AF_INET: 2>,
  <SocketKind.SOCK_STREAM: 1>,
  6,
  '',
  ('127.0.0.1', 22))]
```
- analogical scenario for ipv6:
```
>>> pp(socket.getaddrinfo("::1", 22, 0, socket.SOCK_STREAM))
[(<AddressFamily.AF_INET6: 10>,
  <SocketKind.SOCK_STREAM: 1>,
  6,
  '',
  ('::1', 22, 0, 0))]
```
- for non-resolvable hosts it will return the `socket.gaierror`:
```
>>> socket.getaddrinfo("imaginary.host.address", 22, 0, socket.SOCK_STREAM)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.7/socket.py", line 752, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known
```